src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,671 @@
+                loc21.perform_access(kind, rel2, protected).unwrap();
+                loc21.perform_access(kind, rel1, protected).unwrap();
+
+                // ... and produce the same final result.
+                assert_eq!(
+                    loc12, loc21,
+                    "Read accesses {:?} followed by {:?} do not commute !",
+                    rel1, rel2
+                );
+            }
+        }
+    }
+}
+
+#[test]
+#[rustfmt::skip]
+// Ensure that of 2 accesses happen, one foreign and one a child, and we are protected, that we
+// get UB unless they are both reads.
+fn protected_enforces_noalias() {
+    for [rel1, rel2] in <[AccessRelatedness; 2]>::exhaustive() {
+        // We want to check pairs of accesses where one is foreign and one is not.
+        precondition!(rel1.is_foreign() != rel2.is_foreign());
+        for [kind1, kind2] in <[AccessKind; 2]>::exhaustive() {
+            for mut state in LocationState::exhaustive() {
+                let protected = true;
+                precondition!(state.perform_access(kind1, rel1, protected).is_ok());
+                precondition!(state.perform_access(kind2, rel2, protected).is_ok());
+                // If these were both allowed, it must have been two reads.
+                assert!(
+                    kind1 == AccessKind::Read && kind2 == AccessKind::Read,
+                    "failed to enforce noalias between two accesses that are not both reads"
+                );
+            }
+        }
+    }
+}
+
+/// Possible pairs of `AccessRelatedness` that can occur
+/// to two tags that are mutually foreign.
+/// If there are two pointers `x` and `y` and neither is a child of the other
+/// (which also means they are distinct), then any access is either
+/// - a child of `x` but foreign for `y` (`ChildX`)
+/// - a child of `y` but foreign for `x` (`ChildY`)
+/// - foreign for both (`Other`)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AccessRel2MutuallyForeign {
+    ChildX,
+    ChildY,
+    Other,
+}
+
+impl Exhaustive for AccessRel2MutuallyForeign {
+    fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+        use AccessRel2MutuallyForeign::*;
+        Box::new(vec![ChildX, ChildY, Other].into_iter())
+    }
+}
+
+impl fmt::Display for AccessRel2MutuallyForeign {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use AccessRel2MutuallyForeign::*;
+        match self {
+            ChildX => write!(f, "x"),
+            ChildY => write!(f, "y"),
+            Other => write!(f, "z"),
+        }
+    }
+}
+
+/// Possible pairs of `AccessRelatedness` that can occur
+/// to two tags if one is a child of the other.
+/// If there are two pointers `x` and `y` and `x` is a strict child of `y`,
+/// then any access is either
+/// - a child of both (`ChildBoth`)
+/// - a child of `y` but foreign for `x` (`ChildY`)
+/// - foreign for both (`ChildNeither`)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AccessRel2XChildY {
```

Comment:
This is new, right? Why does it show up now?

Reply:
Yes this is new. While explaining some things above I realized that I didn't prove that this configuration works. It did, but it wasn't tested.